### ISSUE TITLE
`try_cast(parse_json(...))` incorrectly throws error in some cases

### ISSUE THREAD

**Issue by ragnard** [trinodb/trino#21877]
When using `try_cast` on the result of a `parse_json` call, the cast can still fail with an error.

**Reproduction**

```sql
with 
example as (
  select '42' as json
) 
select 
  try_cast(json_parse(json) as array(varchar))
from
  example;
```
```
Query 20240508_195937_00029_qrpb2 failed: Cannot cast to array(varchar). Expected a json array, but got 42
```

From what I can tell, this is due to the optimization "for CAST(JSON_PARSE(...) AS ARRAY/MAP/ROW)" here:

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/sql/planner/IrExpressionInterpreter.java#L769

However, the "safeness" of the original `Cast` has seems to be lost in this process, as the `JsonToXXXFunctions` always error on failure.

**Comment by ragnard** [comment]
I should add this behaviour was introduced in 446. It works as expected in 445 (returns `NULL`).

**Comment by martint** [comment]
Thanks for the report. I'll look into it.

**Comment by martint** [comment]
This is fixed by 0c8a0a681a8c841ade485dbe3b9f6bcb71417ac8 and will be in the next release.
